src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java

```diff
@@ -113,41 +130,145 @@ private void logTargetPath() {
     }
 
+    /**
+     * Sets the odometry position at a given timestamp.
+     *
+     * @param swerveModulePositions the positions of each swerve module
+     * @param gyroHeading           the heading of the gyro
+     * @param timestamp             the timestamp of the odometry observation
+     */
+    private void addOdometryObservation(SwerveModulePosition[] swerveModulePositions, Rotation2d gyroHeading, double timestamp) {
+        updateOdometryPositions(swerveModulePositions, gyroHeading);
+
+        final Twist2d newOdometryPoseDifference = calculateNewOdometryPoseDifference(swerveModulePositions, gyroHeading);
+        updateRobotPosesFromNewOdometryPoseDifference(newOdometryPoseDifference, timestamp);
+    }
+
     private void updateFromVision() {
-        getViableVisionObservations().stream()
-                .sorted(Comparator.comparingDouble(PoseEstimator6328.VisionObservation::timestamp))
-                .forEach(poseEstimator6328::addVisionObservation);
+        updateAllCameras();
+        sortCamerasByLastTargetTimestamp();
+        addVisionObservations();
     }
 
-    private List<PoseEstimator6328.VisionObservation> getViableVisionObservations() {
-        List<PoseEstimator6328.VisionObservation> viableVisionObservations = new ArrayList<>();
+    private void updateAllCameras() {
         for (AprilTagCamera aprilTagCamera : aprilTagCameras) {
-            final PoseEstimator6328.VisionObservation visionObservation = getVisionObservation(aprilTagCamera);
-            if (visionObservation != null)
-                viableVisionObservations.add(visionObservation);
+            aprilTagCamera.update();
         }
-        return viableVisionObservations;
     }
 
-    private PoseEstimator6328.VisionObservation getVisionObservation(AprilTagCamera aprilTagCamera) {
-        aprilTagCamera.update();
-        if (!aprilTagCamera.hasNewResult())
-            return null;
+    private void sortCamerasByLastTargetTimestamp() {
+        VisionObservationSorter.quickSortByDoubleValue(aprilTagCameras, (aprilTagCamera) -> ((AprilTagCamera) aprilTagCamera).getLatestResultTimestampSeconds());
+    }
+
+    /**
+     * Sets the estimated vision pose at the timestamps of the results from the cameras.
+     */
+    private void addVisionObservations() {
+        for (AprilTagCamera aprilTagCamera : aprilTagCameras) {
+            if (!aprilTagCamera.hasNewResult())
+                continue;
+
+            addVisionObservation(
+                    getCameraEstimatedPose(aprilTagCamera),
+                    aprilTagCamera.calculateStandardDeviations(),
+                    aprilTagCamera.getLatestResultTimestampSeconds()
+            );
+        }
+    }
+
+    private Pose2d getCameraEstimatedPose(AprilTagCamera aprilTagCamera) {
         final Pose2d robotPose = aprilTagCamera.getEstimatedRobotPose();
         if (robotPose == null || robotPose.getTranslation() == null || robotPose.getRotation() == null)
             return null;
```

Comment:
no need to null check each part 
which pretty much makes this function un needed

Reply:
With previous logic, the rotation could be set without the translation and would result in an error

Reply:
I don't remember exactly, but it was something to do with the pinhole method returning a value from the gyro but nothing for the actual position of the robot

Reply:
either way, shouldn't exist